### PR TITLE
Rename folder invalidating stat cache

### PR DESCRIPTION
**Pull request by Tulsishah**
### Description
- A function in the lru.go file is needed to erase all entries starting with a given prefix. This is necessary because renameDir is now becoming atomic, and we need to invalidate child folders and directories from the stat cache in case of HNS.
- A function in the stat cache to call the above function from the fast stat bucket.
- Invalidate cache for RenameFolder API from fast stat bucket.
- Haven't integrated stretcher/testify to lru_test as it will add more changes in this PR.

### Link to the issue in case of a bug fix.
NA

### Testing details
1. Manual - NA
2. Unit tests - Added
3. Integration tests - Automated.

### CODE REVIEW

File: internal/cache/metadata/stat_cache_test.go [hunk 1]
```diff
@@ -543,5 +543,51 @@ func (t *StatCacheTest) Test_ShouldEvictEntryOnFullCapacityIncludingFolderSize()
...
+	assert.Nil(t.T(), entry2)
+	assert.False(t.T(), hit3)
+	assert.Nil(t.T(), entry3)
+	assert.False(t.T(), hit4)
+	assert.Nil(t.T(), entry4)
+	assert.True(t.T(), hit5)
+	assert.Equal(t.T(), "b", entry5.Name)
+	assert.True(t.T(), hit6)
+	assert.Equal(t.T(), "d", entry6.Name)
 }
```
Comment (ankitaluthra1) [inline internal/cache/metadata/stat_cache_test.go:593]: Please also add an integration or composite test either in this PR or next to verify that correct entries are deleted with given folder name, current test are not able to verfy what folder names are being passed and if entries have correct prefix corresponding to it.

**Review by ankitaluthra1 — DISMISSED** [review]

File: internal/cache/metadata/stat_cache_test.go (hunk 1 above)
Comment (Tulsishah) [reply internal/cache/metadata/stat_cache_test.go:593]: I'll add e2e tests in upcoming PR for the same.

File: internal/cache/lru/lru.go [hunk 2]
```diff
@@ -267,3 +268,17 @@ func (c *Cache) UpdateWithoutChangingOrder(
 
 	return nil
 }
+
+func (c *Cache) EraseEntriesWithGivenPrefix(
+	prefix string) (value []ValueType) {
```
Comment (vadlakondaswetha) [inline internal/cache/lru/lru.go:273]: Please dont add func arguments in new line. 

Already commented in earlier PR
https://github.com/GoogleCloudPlatform/gcsfuse/pull/2045

File: internal/cache/lru/lru_test.go [hunk 3]
```diff
@@ -169,6 +169,41 @@ func (t *CacheTest) TestEraseWhenKeyPresent() {
 	ExpectEq(nil, t.cache.LookUp("burrito"))
 }
 
+func (t *CacheTest) TestEraseCacheWithGivenPrefix() {
```
Comment (vadlakondaswetha) [inline internal/cache/lru/lru_test.go:172]: Add a testcase when there is no entry matching prefix

File: internal/cache/metadata/stat_cache.go [hunk 4]
```diff
@@ -68,6 +68,15 @@ type StatCache interface {
 	// doesn't exist. Overwrite any existing entry for the name, positive or
 	// negative.
 	AddNegativeEntryForFolder(folderName string, expiration time.Time)
+
+	// Invalidate cache for all the entries with given prefix
+	// e.g. If cache contains below objects
+	// a/b
+	// a/d/c
+	// d
+	// Then it will invalidate entries a, a/b, a/d/c
```
Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache.go:77]: you didnt mention in the comment that cache has object a

File: internal/cache/lru/lru.go [hunk 5]
```diff
@@ -267,3 +268,17 @@ func (c *Cache) UpdateWithoutChangingOrder(
...
+	var deletedEntries []ValueType
+
+	for key := range c.index {
+		if strings.HasPrefix(key, prefix) {
+			e := c.Erase(key)
+			deletedEntries = append(deletedEntries, e)
+		}
+	}
+
+	return deletedEntries
```
Comment (vadlakondaswetha) [inline internal/cache/lru/lru.go:283]: why do you need to returned deletedEntries if its not used>

File: internal/cache/metadata/stat_cache.go [hunk 6]
```diff
@@ -68,6 +68,15 @@ type StatCache interface {
...
+	// Entry d will remain in cache.
+	InValidateCacheForEntriesWithGivenPrefix(prefix string)
```
Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache.go:79]: why not call this method also eraseEntriesWithGivenPrefix

File: internal/cache/metadata/stat_cache_test.go [hunk 7]
```diff
@@ -543,5 +543,51 @@ func (t *StatCacheTest) Test_ShouldEvictEntryOnFullCapacityIncludingFolderSize()
...
+
+	t.statCache.InValidateCacheForEntriesWithGivenPrefix("a")
+
+	hit1, entry1 := t.statCache.LookUpFolder("a", someTime)
+	hit2, entry2 := t.statCache.LookUpFolder("a/b", someTime)
+	hit3, entry3 := t.statCache.LookUp("a/b/c", someTime)
+	hit4, entry4 := t.statCache.LookUpFolder("a/d", someTime)
+	hit5, entry5 := t.statCache.LookUpFolder("b", someTime)
+	hit6, entry6 := t.statCache.LookUp("d", someTime)
+	assert.False(t.T(), hit1)
```
Comment (vadlakondaswetha) [inline internal/cache/metadata/stat_cache_test.go:581]: Ideally you should do assert after each lookup. Proceed to 2nd assertion only if first one fails.

File: internal/storage/caching/fast_stat_bucket.go [hunk 8]
```diff
@@ -328,7 +328,8 @@ func (b *fastStatBucket) GetFolder(
 func (b *fastStatBucket) RenameFolder(ctx context.Context, folderName string, destinationFolderId string) (o *control.RenameFolderOperation, err error) {
 	o, err = b.wrapped.RenameFolder(ctx, folderName, destinationFolderId)
 
-	// TODO: Invalidate cache
+	// Invalidate cache for old directory.
+	b.cache.InValidateCacheForEntriesWithGivenPrefix(folderName)
```
Comment (vadlakondaswetha) [inline internal/storage/caching/fast_stat_bucket.go:332]: why no unit tests for fast_stat_bucket?

File: internal/cache/lru/lru_test.go (hunk 3 above)
Comment (Tulsishah) [reply internal/cache/lru/lru_test.go:172]: Added

File: internal/cache/metadata/stat_cache.go (hunk 4 above)
Comment (Tulsishah) [reply internal/cache/metadata/stat_cache.go:78]: Added

File: internal/cache/metadata/stat_cache_test.go (hunk 7 above)
Comment (Tulsishah) [reply internal/cache/metadata/stat_cache_test.go:576]: Done

File: internal/cache/metadata/stat_cache.go (hunk 6 above)
Comment (Tulsishah) [reply internal/cache/metadata/stat_cache.go]: Done

File: internal/cache/lru/lru.go (hunk 5 above)
Comment (Tulsishah) [reply internal/cache/lru/lru.go]: Erase function is returning deleted entry, so return it here as well. Although laterally found that Erase function entry is getting used other places . So Removed return from here. 

Done.

File: internal/storage/caching/fast_stat_bucket.go (hunk 8 above)
Comment (Tulsishah) [reply internal/storage/caching/fast_stat_bucket.go]: Done

File: internal/cache/lru/lru.go [hunk 9]
```diff
@@ -267,3 +268,11 @@ func (c *Cache) UpdateWithoutChangingOrder(
 
 	return nil
 }
+
+func (c *Cache) EraseEntriesWithGivenPrefix(prefix string) {
+	for key := range c.index {
+		if strings.HasPrefix(key, prefix) {
+			c.Erase(key)
```
Comment (vadlakondaswetha) [inline internal/cache/lru/lru.go:275]: Is it not complaining that you are not handling the return type?

File: internal/storage/caching/fast_stat_bucket_test.go [hunk 10]
```diff
@@ -768,3 +768,18 @@ func (t *StatObjectTest) TestShouldCallGetFolderWhenEntryIsNotPresent() {
 	AssertEq(nil, err)
 	ExpectThat(result, Pointee(DeepEquals(*folder)))
 }
+
+func (t *StatObjectTest) TestShouldRenameFolder() {
```
Comment (vadlakondaswetha) [inline internal/storage/caching/fast_stat_bucket_test.go:772]: fix the test name. 
why there is should in test when you are not specifying any condition

**Review by vadlakondaswetha — APPROVED** [review]

File: internal/cache/lru/lru.go (hunk 9 above)
Comment (Tulsishah) [reply internal/cache/lru/lru.go:275]: It's not throwing any errors. It's only returning deleted entries that we don't need. We're also not getting any warnings about this, so I ignore catching return type.

File: internal/storage/caching/fast_stat_bucket_test.go (hunk 10 above)
Comment (Tulsishah) [reply internal/storage/caching/fast_stat_bucket_test.go]: Done

File: internal/cache/lru/lru.go (hunk 9 above)
Comment (vadlakondaswetha) [reply internal/cache/lru/lru.go:275]: SG

**Review by vadlakondaswetha — APPROVED** [review]

**Pull request merged**